### ISSUE TITLE
the results of torch.inverse is not the same for v1.12.0 and v2.0.1

### ISSUE THREAD

**Issue by 7starsea** [pytorch/pytorch#104158]
### 🐛 Describe the bug

To reproduce the difference of results of ```torch.inverse```, please run the following code using ```v1.12.0``` and ```v2.0.1```
```
import torch
torch.manual_seed(0)
x = torch.rand(10, 10)
x_inv = torch.inverse(x)
print(torch.sum(torch.abs(x_inv)))
```

I also attach another testing code https://github.com/7starsea/pytorch-test, and it seems to me that the result with ```v1.12.0``` is more accurate than the result with ```v2.0.1``` comparing to ```np.linalg.inv```.

## PyTorch version: 2.0.1 env
Collecting environment information...
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Rocky Linux 8.6 (Green Obsidian) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-16)
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.28

Python version: 3.10.9 (main, Mar  1 2023, 18:23:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-425.13.1.el8_7.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 530.30.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              72
On-line CPU(s) list: 0-71
Thread(s) per core:  2
Core(s) per socket:  18
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               106
Model name:          Intel(R) Xeon(R) Gold 6354 CPU @ 3.00GHz
Stepping:            6
CPU MHz:             3600.000
CPU max MHz:         3600.0000
CPU min MHz:         800.0000
BogoMIPS:            6000.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            1280K
L3 cache:            39936K
NUMA node0 CPU(s):   0-17,36-53
NUMA node1 CPU(s):   18-35,54-71
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.3
[pip3] numpydoc==1.5.0
[pip3] pytorch-forecasting==1.0.0
[pip3] pytorch-lightning==2.0.3
[pip3] pytorch-optimizer==2.10.1
[pip3] torch==2.0.1+cu118
[pip3] torch-geometric==2.3.1
[pip3] torchmetrics==0.11.4
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0           py310h7f8727e_0
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0
[conda] mkl_random                1.2.2           py310h00e6091_0
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] numpydoc                  1.5.0           py310h06a4308_0
[conda] pytorch-forecasting       1.0.0                    pypi_0    pypi
[conda] pytorch-lightning         2.0.3                    pypi_0    pypi
[conda] pytorch-optimizer         2.10.1                   pypi_0    pypi
[conda] torch                     2.0.1+cu118              pypi_0    pypi
[conda] torch-geometric           2.3.1                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

## PyTorch version: 1.12.0 env
PyTorch version: 1.12.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Rocky Linux 8.6 (Green Obsidian) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-16)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.28

Python version: 3.8.3 (default, Jul  2 2020, 16:21:59)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-425.13.1.el8_7.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 530.30.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              72
On-line CPU(s) list: 0-71
Thread(s) per core:  2
Core(s) per socket:  18
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               106
Model name:          Intel(R) Xeon(R) Gold 6354 CPU @ 3.00GHz
Stepping:            6
CPU MHz:             3600.000
CPU max MHz:         3600.0000
CPU min MHz:         800.0000
BogoMIPS:            6000.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            1280K
L3 cache:            39936K
NUMA node0 CPU(s):   0-17,36-53
NUMA node1 CPU(s):   18-35,54-71
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==4.0.1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] numpydoc==1.1.0
[pip3] pytorch-lightning==1.7.7
[pip3] torch==1.12.0+cu116
[pip3] torchmetrics==0.11.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.1.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] numpydoc                  1.1.0                      py_0
[conda] pytorch-lightning         1.7.7                    pypi_0    pypi
[conda] torch                     1.12.0+cu116             pypi_0    pypi
[conda] torchmetrics              0.11.0                   pypi_0    pypi

**Comment by malfet** [comment]
Not sure I understand your usecase. `torch.rand` is not guaranteed to yield exactly the values across releases. Also, I'm not sure how printing sum of abs values of inverse matrix should tell one anything about accuracy of the algorithm.
This feels like a more correct test:
```python
import torch
x = torch.rand(10, 10)
x_inv = torch.linalg.inv(x)
print(((x_inv @ x) - torch.eye(10)).abs().max())
```
which at average, should  return something smaller than 1-e5 for float32 data type

**Comment by lezcano** [comment]
@malfet is completely right. The other thing to note is that, if the matrix is singular or close-to-singular, the result may contain larger errors or simply be incorrect if the inverse does not exist. See https://pytorch.org/docs/stable/notes/numerical_accuracy.html#linear-algebra-torch-linalg

**Comment by 7starsea** [comment]
My concern is that the result of ```torch.inverse``` should not be different across releases without any notice. I agree that for testing correctness, the snippet code from @malfet is enough. 

Here is my another simple test
```
import torch
torch.set_printoptions(8)

def testing(device = 'cpu', eps=0.2):
    x = torch.arange(100).reshape(10, 10) + torch.eye(10) * eps
    print(x.dtype, torch.__version__)
    print('det:', torch.linalg.det(x))
    x = x.to(device=device, dtype=torch.float32)
    x_inv = torch.inverse(x)
    print('inv:', torch.sum(torch.abs(x_inv)))
    x_sqrt = torch.sqrt(x)
    print('sqrt:', torch.sum(torch.abs(x_sqrt)))

testing('cpu')
testing('cpu', eps=0.8)
```

Here is the output from torch.2.0.1
```
torch.float32 2.0.1+cu118
det: tensor(-0.02086574)
inv: tensor(89.39737701)
sqrt: tensor(662.05175781)

torch.float32 2.0.1+cu118
det: tensor(-1317.59118652)
inv: tensor(22.49555588)
sqrt: tensor(662.92053223)
```

Here is the output from torch.1.12.0
```
torch.float32 1.12.0+cu116
det: tensor(-0.02086569)
inv: tensor(89.39749908)
sqrt: tensor(662.05175781)

torch.float32 1.12.0+cu116
det: tensor(-1317.58642578)
inv: tensor(22.49557114)
sqrt: tensor(662.92053223)
```
The above tests show that the ```torch.log``` and ```torch.linalg.det``` produce the same results for both v1.12.0 and v2.0.1, but ```torch.inverse``` produces different results. 

Here, my concern is about the difference not the correctness of the api ```torch.inverse``` across release.

As mentioned by @lezcano, if the matrix is singular or close-to-singular, the result may contain larger errors. The larger the absolute value of determinant, the smaller the difference for the inverse.

**Comment by lezcano** [comment]
Alas, this is expected and within tolerance range. One should not expect the results to be bitwise equivalent between releases. PyTorch cannot guarantee this. The relevant documentation is in https://pytorch.org/docs/stable/notes/numerical_accuracy.html

I'm happy to continue the discussion in the forums https://discuss.pytorch.org/
